Nodetool additional commands 4/N

denesb opened the pull request:
This PR implements the following new nodetool commands:
* snapshot
* drain
* flush
* disableautocompaction
* enableautocompaction

All commands come with tests and all tests pass with both the new and the current nodetool implementations.

Refs: https://github.com/scylladb/scylladb/issues/15588

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - Build
:white_check_mark: - Sanity Tests
:white_check_mark: - Unit Tests
#### Build Details:
- Build URL: https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/4509/
- Duration: 2 hr 7 min
- Builder: spider4.cloudius-systems.com

nyh commented on test/nodetool/README.md:
````diff
@@ -0,0 +1,23 @@
+```
+
+Again, you can specify the path to the nodetool binary with `--nodetool-path` option. By default, `scylladb.git/tools/java/bin/nodetool` will be used.
+You can specify the path to JMX with `--jmx-path` option. By default, `scylladb.git/tools/jmx/scripts/scylla-jmx` will be used.
````
Can you please explain in more detail what JMX is used for? I assume it's only needed for the "--nodetool=cassandra" case, and not needed for the Scylla-native nodetool?

denesb replied:
Yes, therefore I only mention `JMX` here at the section about `--nodetool=cassandra`.

nyh replied:
Ok, I didn't understand it was inside some specific section.

denesb replied:
Added further clarifications in v2.

nyh reviewed (approved):
Looks good. I had some small comments, but you can consider them for followups.

denesb commented on test/nodetool/test_snapshot.py:
```diff
@@ -99,3 +102,195 @@ def test_listsnapshots(nodetool, request):
+
+def check_snapshot_out(res, tag, ktlist, skip_flush):
+    if type(tag) is int:
+        tag_pattern = str(tag)[:-5] + "[0-9]{5}"
```
I thought it is implied, I omit `nodetool` from names because we are in `test/nodetool` directory here. :)

nyh replied:
This is a utility function that does something *inside* this file. It checks the standard output of the "nodetool snapshot" command - this can either be explained in a comment or a good name. As it stands, the name "check_snapshot_out" led me to misunderstand it checks the actual snapshot. Which isn't even possible because everything in this test is mocked and no actual snapshot is ever created.
Anyway, it was just a thought. You don't have to change it.

denesb replied:
Added comment for the method in v2. Also, this tricky tag matching is much clearer in v2.

denesb commented on test/nodetool/test_snapshot.py:
```diff
@@ -99,3 +102,195 @@ def test_listsnapshots(nodetool, request):
+    {"ks": ["ks1.tbl1", "ks2.tbl2"]},
+    {"ks": ["ks1"], "tbl": ["tbl1"]},
+    {"ks": ["ks1"], "tbl": ["tbl1", "tbl2"]},
+    {"ks": ["ks1", "ks2"], "tbl": []},
```
I'm only checking things that nodetool itself validates here. For now, I remember a single command in nodetool having special casing set aside for alternator. I think if we started checking this systematically, we would find that a lot of commands are broken by such names today.

nyh replied:
Sure. We can do this later. One day we should.

denesb commented on test/nodetool/README.md:
````diff
@@ -0,0 +1,23 @@
+pytest --nodetool=cassandra .
+```
+
+Again, you can specify the path to the nodetool binary with `--nodetool-path` option. By default, `scylladb.git/tools/java/bin/nodetool` will be used.
````
`scylladb.git` stands for the root directory of ... `scylladb.git`. I didn't want to use relative paths, because this README.md is in the `tests/nodetool` directory and I don't know how the reader would interpret relative paths.

nyh replied:
Ok.

denesb replied:
I changed this to `<scylladb.git>` so it is more clear, I'm talking about the root of the repo.

tchaikov commented on test/nodetool/test_snapshot.py:
```diff
@@ -99,3 +102,195 @@ def test_listsnapshots(nodetool, request):
+
+def check_snapshot_out(res, tag, ktlist, skip_flush):
+    if type(tag) is int:
+        tag_pattern = str(tag)[:-5] + "[0-9]{5}"
```
this introduces a flaky test. here is a real world example: tag = `1699251599822`. and after a while, we would have a snapshot whose tag is `1699251600039`, please note
`1699251600039-1699251599822` is just 217.

see https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/4530/testReport/junit/test.nodetool/test_snapshot/Tests___Unit_Tests___test_snapshot_options_matrix_False_ktlist5_None_/ and https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/4530/testReport/junit/test.nodetool/test_snapshot/Tests___Unit_Tests___test_snapshot_options_matrix_False_ktlist5_None__2/

tchaikov replied:
```
AssertionError: assert None
 +  where None = <built-in method fullmatch of re.Pattern object at 0x5644cc4d8e00>('Requested creating snapshot(s) for [ks1,ks2] with snapshot name [1699251600039] and options {skipFlush=false}')
 +    where <built-in method fullmatch of re.Pattern object at 0x5644cc4d8e00> = re.compile('Requested creating snapshot\\(s\\) for \\[ks1, ?ks2\\] with snapshot name \\[16992515[0-9]{5}\\] and options \\{skipFlush=false\\}').fullmatch
..
nodetool = <function nodetool.<locals>.invoker at 0x7f6c8386a5c0>
tag = 1699251599822, ktlist = {'ks': ['ks1', 'ks2'], 'tbl': []}
skip_flush = False

    @pytest.mark.parametrize("tag", [None, "my_snapshot_tag"])
    @pytest.mark.parametrize("ktlist", [
        {},
        {"ks": ["ks.tbl1"]},
        {"ks": ["ks1.tbl1", "ks2.tbl2"]},
        {"ks": ["ks1"], "tbl": ["tbl1"]},
        {"ks": ["ks1"], "tbl": ["tbl1", "tbl2"]},
        {"ks": ["ks1", "ks2"], "tbl": []},
    ])
    @pytest.mark.parametrize("skip_flush", [False, True])
    def test_snapshot_options_matrix(nodetool, tag, ktlist, skip_flush):
        cmd = ["snapshot"]
        params = {}
    
        if tag is None:
            tag = int(time.time() * 1000)
            params["tag"] = approximate_value(value=tag, delta=99999).to_json()
        else:
            cmd += ["--tag", tag]
            params["tag"] = tag
    
        if skip_flush:
            cmd.append("--skip-flush")
    
        params["sf"] = str(skip_flush).lower()
    
        if ktlist:
            if "tbl" in ktlist:
                if len(ktlist["tbl"]) > 0:
                    keyspaces = ktlist["ks"]
                    cmd += ["--table", ",".join(ktlist["tbl"])]
                    cmd += keyspaces
                    params["kn"] = keyspaces[0]
                    params["cf"] = ",".join(ktlist["tbl"])
                else:
                    keyspaces = ktlist["ks"]
                    cmd += keyspaces
                    params["kn"] = ",".join(keyspaces)
            else:
                keyspaces = ktlist["ks"]
                cmd += ["-kt", ",".join(keyspaces)]
                if len(keyspaces) == 1:
                    ks, tbl = keyspaces[0].split(".")
                    params["kn"] = ks
                    params["cf"] = tbl
                elif len(keyspaces) > 1:
                    params["kn"] = ",".join(keyspaces)
        else:
            keyspaces = []
    
        res = nodetool(*cmd, expected_requests=[
            expected_request("POST", "/storage_service/snapshots", params=params)
        ])
    
>       check_snapshot_out(res, tag, keyspaces, skip_flush)

test/nodetool/test_snapshot.py:238: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

res = 'Requested creating snapshot(s) for [ks1,ks2] with snapshot name [1699251600039] and options {skipFlush=false}\nSnapshot directory: 1699251600039\n'
tag = 1699251599822, ktlist = ['ks1', 'ks2'], skip_flush = False

    def check_snapshot_out(res, tag, ktlist, skip_flush):
        if type(tag) is int:
            tag_pattern = str(tag)[:-5] + "[0-9]{5}"
        else:
            tag_pattern = tag
    
        if len(ktlist) == 0:
            keyspaces = "all keyspaces"
        else:
            keyspaces = ", ?".join(ktlist)
    
        main_pattern = re.compile("Requested creating snapshot\\(s\\)"
                                  f" for \\[{keyspaces}\\]"
                                  f" with snapshot name \\[{tag_pattern}\\]"
                                  f" and options \\{{skipFlush={str(skip_flush).lower()}\\}}")
    
        snapshot_dir_pattern = re.compile(f"Snapshot directory: {tag_pattern}")
    
        print(res)
    
        lines = res.split("\n")
        assert len(lines) == 3
>       assert main_pattern.fullmatch(lines[0])
E       AssertionError: assert None
E        +  where None = <built-in method fullmatch of re.Pattern object at 0x5644cc4d8e00>('Requested creating snapshot(s) for [ks1,ks2] with snapshot name [1699251600039] and options {skipFlush=false}')
E        +    where <built-in method fullmatch of re.Pattern object at 0x5644cc4d8e00> = re.compile('Requested creating snapshot\\(s\\) for \\[ks1, ?ks2\\] with snapshot name \\[16992515[0-9]{5}\\] and options \\{skipFlush=false\\}').fullmatch

test/nodetool/test_snapshot.py:129: AssertionError
```

nyh replied:
Wow, I knew I didn't understand what this trick was trying to do (I said: "I have no idea what this weird construct does, or actually what this function wants to do or what its "tag" parameter actually is. ) but I missed that it's a downright bug.
@denesb I'll dequeue the PR.

tchaikov replied:
probably we can use the homebrew `approximate_value` helper or pytest.approx for testing this? but i'd still like to understand what the integer represented by "tag" is for and why we use "5" here. probably because its a timestamp in microsecond?

denesb replied:
Fixed in v2. I now extract the tag and match it according to the type.

tchaikov commented on test/nodetool/rest_api_mock.py:
```diff
@@ -19,6 +19,19 @@
 logger = logging.getLogger(__name__)
 
 
+class approximate_value:
```
i think pytest already provides [pytest.approx](https://docs.pytest.org/en/latest/reference/reference.html#pytest.approx). does it offer us what we need here?

denesb replied:
This doesn't work for me, I have to send this type over the wire.

tchaikov reviewed (changes requested):
the test failures in #15691 and #15958 led me here.

nyh commented:
> the test failures in #15691 and #15958 led me here.

Can you please explain how it led you here?
For example, in https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/4444/testReport/junit/compaction_additional_test/TestCompactionAdditional/Tests___Sanity_Tests___test_compaction_delete_with_smp_change/ we have "nodetool compact" (the old, traditional, nodetool - not Botond's new work), causing a Scylla crash. How is this related in any way to this PR?

tchaikov commented:
> > the test failures in #15691 and #15958 led me here.
> 
> Can you please explain how it led you here? For example, in https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/4444/testReport/junit/compaction_additional_test/TestCompactionAdditional/Tests___Sanity_Tests___test_compaction_delete_with_smp_change/ we have "nodetool compact" (the old, traditional, nodetool - not Botond's new work), causing a Scylla crash. How is this related in any way to this PR?

@nyh hi Nadav, please see https://github.com/scylladb/scylladb/pull/15939#discussion_r1382878761 . please note the jenkins picks a random junit test result, so there are chances i cannot get the failure one. but i copied the failure in the comment for the reference.

nyh commented:
Dequeued this pull request due to the bug discovered by @tchaikov above that led to test failure.
@denesb please send a new fixed version for this pull request. Thanks.

nyh commented:
> @nyh hi Nadav, please see [#15939 (comment)](https://github.com/scylladb/scylladb/pull/15939#discussion_r1382878761) . please also note that jenkins picks a random junit test result, so there are chances we cannot get the failure one following the link. but i copied the failure in the comment for the reference.

Thanks. You're right, I missed that one of the failures was caused by this PR, and dequeued this PR now. It's sad that we have other test failures as well in that run, that probably means we have other bugs we need to fix for CI stability.

denesb commented:
v2:
* Rebased
* Fixed used-after-free discovered on ARM
* Fix snapshot tag matching: instead of matching the tag as a string, extract it and match it separately, taking into account the type of the tag.

denesb commented on test/nodetool/rest_api_mock.py:
```diff
@@ -19,6 +19,19 @@
 logger = logging.getLogger(__name__)
 
 
+class approximate_value:
```
Added comment in v2.

denesb commented:
Hold on with merging. I have one more fix to add.

denesb commented:
v3:
* fix {enable,disable}autocompaction to act on all keyspaces when there are no args

nyh commented on test/nodetool/test_snapshot.py:
```diff
@@ -99,3 +102,201 @@ def test_listsnapshots(nodetool, request):
+    if type(tag) is int:
+        # The tag is a timestamp in millis.
+        # Allow for 10s worth of divergence between actual and expected tag value.
+        assert int(actual_tag) - tag <= 10000
```
You don't have here abs(), so if the difference is negative, it can be very very negative and it's fine. Is this intentional?

nyh replied:
By the way, given the crazy-slow-tests we see once in a while, is 10 seconds enough? :-)

denesb replied:
In practice `actual` will always be larger than `tag`, because `tag` is generated by the test and `actual_tag` is generated by nodetool, at a later point in time. But just to be on the safe size, I will increase this.
I will also increase the allowed difference to 30sec.

nyh replied:
Yes, in practice this test will always pass because nodetool really does what we expect it to do ;-) But isn't the point to check that it does?

nyh reviewed (approved):
Looks good to me, I just have a question about a missing (?) abs().

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - Build
:white_check_mark: - Sanity Tests
:white_check_mark: - Unit Tests
#### Build Details:
- Build URL: https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/4571/
- Duration: 2 hr 8 min
- Builder: spider3.cloudius-systems.com

denesb commented:
v3:
* Use abs() and increase tolerance to 30s when checking snapshot tag timestamp.

denesb commented:
BTW, with this PR, there are now in total 106 tests.
With a ScyllaDB dev build, it takes 21.65s to run all these in a sequential manner.
The same takes 96.26s with the Java nodetool (which even skips 12 tests).

So even with all the seastar machinery, which is massive overkill for a simple tool such as nodetool, the scylla-native nodetool is still > 5X faster than the Java-based one. Looks like the JVM is even more overkill for a simple tool, than seastar.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - Build
:white_check_mark: - Sanity Tests
:x: - Unit Tests
#### Failed Tests (4/17232):
- [test_tablet_metadata_propagates_with_schema_changes_in_snapshot_mode](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/4580/testReport/junit/%28root%29/test_tablets/Tests___Unit_Tests___test_tablet_metadata_propagates_with_schema_changes_in_snapshot_mode) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tablet_metadata_propagates_with_schema_changes_in_snapshot_mode)
- [test_scans](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/4580/testReport/junit/%28root%29/test_tablets/Tests___Unit_Tests___test_scans) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_scans)
- [test_table_drop_with_auto_snapshot](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/4580/testReport/junit/%28root%29/test_tablets/Tests___Unit_Tests___test_table_drop_with_auto_snapshot) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_table_drop_with_auto_snapshot)
- [test_topology_changes](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/4580/testReport/junit/%28root%29/test_tablets/Tests___Unit_Tests___test_topology_changes) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_topology_changes)
#### Build Details:
- Build URL: https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/4580/
- Duration: 2 hr 25 min
- Builder: spider5.cloudius-systems.com

avikivity commented:
> BTW, with this PR, there are now in total 106 tests. With a ScyllaDB dev build, it takes 21.65s to run all these in a sequential manner. The same takes 96.26s with the Java nodetool (which even skips 12 tests).
> 
> So even with all the seastar machinery, which is massive overkill for a simple tool such as nodetool, the scylla-native nodetool is still > 5X faster than the Java-based one. Looks like the JVM is even more overkill for a simple tool, than seastar.

Seastar overhead isn't high, I'm guessing most of the time is spent on the server side.

nyh commented:
> BTW, with this PR, there are now in total 106 tests. With a ScyllaDB dev build, it takes 21.65s to run all these in a sequential manner. The same takes 96.26s with the Java nodetool (which even skips 12 tests).
> 
> So even with all the seastar machinery, which is massive overkill for a simple tool such as nodetool, the scylla-native nodetool is still > 5X faster than the Java-based one. Looks like the JVM is even more overkill for a simple tool, than seastar.

Yes, when I was working in IBM and we used Java a lot, I used to joke that everybody loves to program in Java, but once you have the end result, everybody is sorry it was written in Java :-( Short-running utilities written in Java took ridiculous (relatively...) amount of time to start, and long-running utilities written in Java took ridiculous amounts of memory. The only Java products that didn't suck were server products which were anyway designed to take over the entire server and all its memory - sort of like what Seastar does too.

21 seconds for 106 tests is 0.2 seconds per test. Since all these tests basically just run the nodetool executable once, I guess it means it takes 0.2 seconds to run this tool. This is not terrible, but still unfortunate - it's almost as slow as booting a full-blown Scylla (0.3 seconds on my laptop), and I wonder if it can be improved. But I guess we don't need to run to improve it right now - it's not _that_ bad :-)

nyh commented:
> Seastar overhead isn't high

Maybe it's the startup time. Maybe Seastar is starting and allocating tons of memory (can you explicitly run it with only a bit of memory)? Or maybe the executable is huge and it takes time to process it, dynamically link it, or something. But these are just guesses.

> I'm guessing most of the time is spent on the server side.

Interesting guess: In his tests, the server is a mock which just receives the request and immediately returns a canned response. It shouldn't be slow. But shouldn't doesn't mean it isn't! A common bug causing frustratingly high latencies in HTTP server is Nagle's algorithm + delayed ack - e.g., https://github.com/scylladb/seastar/issues/968. Maybe your HTTP server has this problem?

denesb commented:
> > Seastar overhead isn't high
> 
> Maybe it's the startup time. Maybe Seastar is starting and allocating tons of memory (can you explicitly run it with only a bit of memory)? Or maybe the executable is huge and it takes time to process it, dynamically link it, or something. But these are just guesses.

Our tools use the standard allocator and preallocate only a small amount (1 MiB) of memory. Tools also disable all the optimizations that are only useful when seastar takes over the entire machine -- which is not the case for tools.

> > I'm guessing most of the time is spent on the server side.
> 
> Interesting guess: In his tests, the server is a mock which just receives the request and immediately returns a canned response. It shouldn't be slow. But shouldn't doesn't mean it isn't! A common bug causing frustratingly high latencies in HTTP server is Nagle's algorithm + delayed ack - e.g., [scylladb/seastar#968](https://github.com/scylladb/seastar/issues/968). Maybe your HTTP server has this problem?

The server side is the same so if that was slow, it would affect both sides in equal manner. Note that the Java based nodetool has an additional hop -- JMX. This certainly contributes but the main contributor I think is JVM startup time.

denesb commented:
> #### 🔴 CI State: FAILURE
> 
> ✅ - Build ✅ - Sanity Tests ❌ - Unit Tests
> #### Failed Tests (4/17232):
> 
>     * [test_tablet_metadata_propagates_with_schema_changes_in_snapshot_mode](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/4580/testReport/junit/%28root%29/test_tablets/Tests___Unit_Tests___test_tablet_metadata_propagates_with_schema_changes_in_snapshot_mode) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tablet_metadata_propagates_with_schema_changes_in_snapshot_mode)
> 
>     * [test_scans](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/4580/testReport/junit/%28root%29/test_tablets/Tests___Unit_Tests___test_scans) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_scans)
> 
>     * [test_table_drop_with_auto_snapshot](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/4580/testReport/junit/%28root%29/test_tablets/Tests___Unit_Tests___test_table_drop_with_auto_snapshot) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_table_drop_with_auto_snapshot)
> 
>     * [test_topology_changes](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/4580/testReport/junit/%28root%29/test_tablets/Tests___Unit_Tests___test_topology_changes) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_topology_changes)
> 
> 
> #### Build Details:
> 
>     * Build URL: https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/4580/
> 
>     * Duration: 2 hr 25 min
> 
>     * Builder: spider5.cloudius-systems.com

https://github.com/scylladb/scylladb/issues/15991

nyh commented:
> Our tools use the standard allocator and preallocate only a small amount (1 MiB) of memory. Tools also disable all the optimizations that are only useful when seastar takes over the entire machine -- which is not the case for tools.

Well, something is taking an average of 0.2 seconds per test. What is it?
Maybe I'm misunderstanding and on average each test does many "nodetool" runs, not one?

> The server side is the same so if that was slow, it would affect both sides in equal manner. Note that the Java based nodetool has an additional hop -- JMX. This certainly contributes but the main contributor I think is JVM startup time.

Of course, the Java one is very slow, I'm not even thinking about that. I'm just thinking why the new code is "slow" (0.2 seconds per test). My guess was that maybe we're relatively getting high latency (40ms) for each mocked REST API request, and that adds up. But it's just a guess. Something needs to add up to 0.2 seconds...

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - Build
:white_check_mark: - Sanity Tests
:white_check_mark: - Unit Tests
#### Build Details:
- Build URL: https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/4595/
- Duration: 2 hr 7 min
- Builder: spider2.cloudius-systems.com

denesb commented:
> Well, something is taking an average of 0.2 seconds per test. What is it?
> Maybe I'm misunderstanding and on average each test does many "nodetool" runs, not one?

On average, each test runs nodetool once, but nodetool will on average do multiple (2-3) API calls in a single call.
I will at one point do a measurement and extract the run-down of what is taking how much time.
